{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a99240b3",
   "metadata": {},
   "source": [
    "## Reproduce Figure 5b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db325144",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import ipynbname\n",
    "REPO_PATH = ipynbname.path().parent.parent"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1870f70",
   "metadata": {},
   "source": [
    "## Launch predictions using Variantformer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0453ccd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Essential imports\n",
    "import sys\n",
    "from pathlib import Path\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "\n",
    "from processors.variantprocessor import VariantProcessor"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96b29a32",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37450a7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants = pd.read_parquet(os.path.join(REPO_PATH, '_artifacts/b23c9b69.pq'))\n",
    "variants['tissue'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28dbccb3",
   "metadata": {},
   "source": [
    "### Predict for first 10 variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d6c4f98",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tiisue map to align tissue names between VF and Eqtl catalog data\n",
    "vf_tissue_map = {'skin' : 'skin - sun exposed (lower leg)',\n",
    "                 'blood' : 'whole blood',\n",
    "                 'adipose' : 'adipose - subcutaneous',\n",
    "                 'brain - frontal cortex (ba9)': 'brain - frontal cortex (ba9)',\n",
    "                 'brain - putamen (basal ganglia)': 'brain - putamen (basal ganglia)',\n",
    "                 'brain - substantia nigra' : 'brain - substantia nigra'\n",
    "                 }\n",
    "variants['tissue'] = variants['tissue'].map(vf_tissue_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71fc07af",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "first10_variants = variants.head(10)\n",
    "first10_variants['chr'] = first10_variants['variant_id'].apply(lambda x: x.split('_')[0])\n",
    "first10_variants['pos'] = first10_variants['variant_id'].apply(lambda x: int(x.split('_')[1]))\n",
    "first10_variants['ref'] = first10_variants['variant_id'].apply(lambda x: x.split('_')[2])\n",
    "first10_variants['alt'] = first10_variants['variant_id'].apply(lambda x: x.split('_')[3])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ca44c3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants['tissue'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73844833",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize VariantFormer\n",
    "print(\"ðŸš€ Initializing VariantFormer Variant Processor...\")\n",
    "model_class = 'v4_ag' # model class can be 'v4_ag', 'v4_pcg'. AG model is all-genes model trained on both protein-coding and non-coding genes.\n",
    "# model_class = 'v4_pcg' # Uncomment to use the PCG model\n",
    "vep = VariantProcessor(model_class=model_class)\n",
    "# Run variant predictions\n",
    "print(\"ðŸ”¬ Running VariantFormer variant analysis...\")\n",
    "output_dir = \"/tmp/vep_eqtl_output\"\n",
    "# Predict expression effects for all variants\n",
    "raw_predictions = vep.predict(var_df=first10_variants, output_dir=output_dir)\n",
    "print(\"Formatting VariantFormer scores and computing eQTL statistics...\")\n",
    "formatted_scores = vep.format_scores(raw_predictions)\n",
    "print(\"Computing eQTL statistics...\")\n",
    "final_results = vep.eqtl_scores(formatted_scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c2351f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_results.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "677df43c",
   "metadata": {},
   "source": [
    "## Launch predictions using Alphagenome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "747b1f54",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "489154a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants = pd.read_parquet(os.path.join(REPO_PATH, '_artifacts/b23c9b69.pq'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99e32323",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tiisue map to align tissue names between VF and Eqtl catalog data\n",
    "alphagenome_tissue_map = {'skin' : 'Skin_Sun_Exposed_Lower_leg',\n",
    "                 'blood' : 'Whole_Blood',\n",
    "                 'adipose' : 'Adipose_Subcutaneous',\n",
    "                 'brain - frontal cortex (ba9)': 'Brain_Frontal_Cortex_BA9',\n",
    "                 'brain - putamen (basal ganglia)': 'Brain_Putamen_basal_ganglia',\n",
    "                 'brain - substantia nigra' : 'Brain_Substantia_nigra'\n",
    "                 }\n",
    "variants['tissue'] = variants['tissue'].map(alphagenome_tissue_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64b7e620",
   "metadata": {},
   "outputs": [],
   "source": [
    "first10_variants = variants.head(10)\n",
    "first10_variants['CHROM'] = first10_variants['variant_id'].apply(lambda x: x.split('_')[0])\n",
    "first10_variants['POS'] = first10_variants['variant_id'].apply(lambda x: int(x.split('_')[1]))\n",
    "first10_variants['REF'] = first10_variants['variant_id'].apply(lambda x: x.split('_')[2])\n",
    "first10_variants['ALT'] = first10_variants['variant_id'].apply(lambda x: x.split('_')[3])\n",
    "first10_variants['gene_id'] = first10_variants['gene_id'].apply(lambda x: x.split('.')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32b03eba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install alphagenome\n",
    "!uv pip install alphagenome\n",
    "from io import StringIO\n",
    "from alphagenome import colab_utils\n",
    "from alphagenome.data import genome\n",
    "from alphagenome.models import dna_client, variant_scorers\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf62518b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the model.\n",
    "API_KEY = \"your_api_key_here\"\n",
    "dna_model = dna_client.create(API_KEY)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e10d15e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Alphagenome specifications\n",
    "organism = 'human'\n",
    "# @markdown Specify length of sequence around variants to predict:\n",
    "sequence_length = '1MB'\n",
    "sequence_length = dna_client.SUPPORTED_SEQUENCE_LENGTHS[\n",
    "    f'SEQUENCE_LENGTH_{sequence_length}'\n",
    "]\n",
    "score_rna_seq = True\n",
    "download_predictions = False\n",
    "# Parse organism specification.\n",
    "organism_map = {\n",
    "    'human': dna_client.Organism.HOMO_SAPIENS,\n",
    "    'mouse': dna_client.Organism.MUS_MUSCULUS,\n",
    "}\n",
    "organism = organism_map[organism]\n",
    "# Parse scorer specification.\n",
    "scorer_selections = {\n",
    "    'rna_seq': score_rna_seq,\n",
    "}\n",
    "all_scorers = variant_scorers.RECOMMENDED_VARIANT_SCORERS\n",
    "selected_scorers = [\n",
    "    all_scorers[key]\n",
    "    for key in all_scorers\n",
    "    if scorer_selections.get(key.lower(), False)\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87bae8b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = []\n",
    "for i, vcf_row in first10_variants.iterrows():\n",
    "  variant = genome.Variant(\n",
    "      chromosome=str(vcf_row.CHROM),\n",
    "      position=int(vcf_row.POS),\n",
    "      reference_bases=vcf_row.REF,\n",
    "      alternate_bases=vcf_row.ALT,\n",
    "      name=vcf_row.variant_id,\n",
    "  )\n",
    "  interval = variant.reference_interval.resize(sequence_length)\n",
    "\n",
    "  variant_scores = dna_model.score_variant(\n",
    "      interval=interval,\n",
    "      variant=variant,\n",
    "      variant_scorers=selected_scorers,\n",
    "      organism=organism,\n",
    "  )\n",
    "  df_scores = variant_scorers.tidy_scores([variant_scores])\n",
    "  tissue_key = vcf_row.tissue\n",
    "  df = df_scores[(df_scores['gene_id'] == vcf_row.gene_id) & (df_scores['gtex_tissue'] == tissue_key)]['raw_score']\n",
    "  vcf_row['alphagenome_score_new'] = df.values[0].mean() if not df.empty else np.nan\n",
    "  results.append(vcf_row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4afb797",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82341c60",
   "metadata": {},
   "source": [
    "**The alphagenome score might deviate a little because we believe the underlying model is evolving** "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2029fe2f",
   "metadata": {},
   "source": [
    "## Visualize all the precalculated scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "747f1f32",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants = pd.read_parquet(os.path.join(REPO_PATH, '_artifacts/b23c9b69.pq'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6ae6fba",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d5442ac",
   "metadata": {},
   "source": [
    "### VF results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b719f41",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tissue specific spoearman correlation plots\n",
    "vf_spearman = variants.groupby('tissue').apply(\n",
    "    lambda x: x['slope'].corr(x['VF-agg-log2fc-weighted_ag'], method='spearman')\n",
    ").reset_index()\n",
    "vf_spearman.columns = ['tissue', 'spearman_correlation']\n",
    "vf_spearman\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77355fe9",
   "metadata": {},
   "source": [
    "### Alphagenome results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44f0e516",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tissue specific spoearman correlation plots\n",
    "alphagenome_spearman = variants.groupby('tissue').apply(\n",
    "    lambda x: x['slope'].corr(x['alphagenome_score'], method='spearman')\n",
    ").reset_index()\n",
    "alphagenome_spearman.columns = ['tissue', 'spearman_correlation']\n",
    "alphagenome_spearman"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a53fc24",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
