{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d185ec07",
   "metadata": {},
   "source": [
    "# Insilico mutation effect of Alzheimer's risk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62556c70",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Essential imports\n",
    "import sys\n",
    "from pathlib import Path\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "import ipynbname\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "\n",
    "from processors.variantprocessor import VariantProcessor\n",
    "REPO_PATH = ipynbname.path().parent.parent"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0718a44",
   "metadata": {},
   "source": [
    "## Setup VariantProcessor to generate embeddings\n",
    "**model_class** : `v4_ag` or `v4_pcg`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a13c1fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_class = 'v4_pcg' # model class can be 'v4_ag', 'v4_pcg'. AG model is all-genes model trained on both protein-coding and non-coding genes.\n",
    "# model_class = 'v4_pcg' # Uncomment to use the PCG model\n",
    "vep = VariantProcessor(model_class=model_class)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d87c8331",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example VCF integration parameters\n",
    "vcf_path = os.path.join(REPO_PATH,\"_artifacts/HG00096.vcf.gz\") # Change path as needed\n",
    "sample_name = \"HG00096\"  # European sample from 1000 Genomes Project\n",
    "\n",
    "# Same variant for insilico analysis. Alt allele will be placed in the sample genotype context at that specific position. \n",
    "# Ref is from reference genome.\n",
    "vcf_variant_data = {\n",
    "    \"chr\": [\"chr19\"],\n",
    "    \"pos\": [44908684],\n",
    "    \"ref\": [\"T\"], # reference allele in hg38\n",
    "    \"alt\": [\"C\"], # insilico edited allele\n",
    "    \"tissue\": [\"brain - cortex\"],\n",
    "    \"gene_id\": [\"ENSG00000130203.9\"],\n",
    "}\n",
    "\n",
    "vcf_variant_df = pd.DataFrame(vcf_variant_data)\n",
    "print(\"\\nüß¨ Analyzing the same variant with individual genotype data:\")\n",
    "print(vcf_variant_df.to_string(index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13023517",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run VariantFormer with VCF integration\n",
    "print(\"\\nüî¨ Running VariantFormer analysis with VCF integration...\")\n",
    "print(\"‚è≥ Processing individual genotype data...\")\n",
    "output_dir = \"/tmp/variant_risk_output\"  # Change output directory as needed\n",
    "vcf_predictions = vep.predict(\n",
    "    var_df=vcf_variant_df,\n",
    "    output_dir=output_dir,\n",
    "    vcf_path=vcf_path,\n",
    "    sample_name=sample_name,\n",
    ")\n",
    "\n",
    "print(\"‚úÖ VCF-based predictions completed!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f764d9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_predictions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10df2f08",
   "metadata": {},
   "source": [
    "**Prediction Output Schema**\n",
    "- `chrom`, `pos`, `ref`, `alt`: Variant coordinates and alleles\n",
    "- `genes`: Ensembl Gene ID\n",
    "- `tissue`: Tissue name\n",
    "- `population`: Population code (REF_HG38, EAS, EUR, AFR, SAS, AMR)\n",
    "- `sample_name`: 1000 Genomes sample identifier\n",
    "- `zygosity`: 0 (homozygous ref), 1 (heterozygous), 2 (homozygous alt)\n",
    "- `gene_exp`: Predicted gene expression level for the gene in the specified tissue and sample\n",
    "- `variant_type`: Gene Overlap / CRE Overlap\n",
    "- `gene_emb`: Embedding of the gene conditioned on tissue and regulatory context from the last transformer layer\n",
    "- `gene_token_embedding`:  Embedding of the token that overlaps with the variant position from the last transformer layer\n",
    "- `cre_token_embedding`: Embeddings of regulatory elements overlapping with the variant position from the last transformer layer\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b795dd12",
   "metadata": {},
   "source": [
    "## Predict Adrisk with and without insilico edits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "583270d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from processors import ad_risk"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5a85514",
   "metadata": {},
   "source": [
    "### Subselect respective rows that corresponds to the edits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03f76323",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_embedding_without_insilico_edit = vcf_predictions[(vcf_predictions['zygosity']=='0')& (vcf_predictions['sample_name']==sample_name) ]\n",
    "gene_embedding_with_homozygous_insilico_edit = vcf_predictions[(vcf_predictions['zygosity']=='2')& (vcf_predictions['sample_name']==sample_name) ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99f14a1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load tissue vocab\n",
    "tissue_vocab = vep.tissue_vocab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98b80674",
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in gene_embedding_without_insilico_edit.itertuples():\n",
    "    gene_id = row.genes\n",
    "    tissue_id = tissue_vocab[row.tissues]\n",
    "    embd = row.gene_emb.reshape(1, -1)\n",
    "    risk = ad_risk.ADrisk(\n",
    "        model_class=model_class,\n",
    "        gene_id=gene_id,\n",
    "        tissue_id=tissue_id)(embd)\n",
    "    print(f\"Gene: {gene_id}, Tissue: {row.tissues}, Risk without insilico edit: {risk[0]:.4f}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78aa8bcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in gene_embedding_with_homozygous_insilico_edit.itertuples():\n",
    "    gene_id = row.genes\n",
    "    tissue_id = tissue_vocab[row.tissues]\n",
    "    embd = row.gene_emb.reshape(1, -1)\n",
    "    risk = ad_risk.ADrisk(\n",
    "        model_class=model_class,\n",
    "        gene_id=gene_id,\n",
    "        tissue_id=tissue_id)(embd)\n",
    "    print(f\"Gene: {gene_id}, Tissue: {row.tissues}, Risk with homozygous insilico edit: {risk[0]:.4f}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
